src/coalescenceml/orchestrator/utils.py

```diff
@@ -0,0 +1,282 @@
+    logger.info(
+        f"Step `{pipeline_step_name}` has finished in "
+        f"{readability_utils.get_human_readable_time(run_duration)}."
+    )
+    return execution_info
+
+
+def get_step_for_node(node: PipelineNode, steps: List[BaseStep]) -> BaseStep:
+    """Finds the matching step for a tfx pipeline node."""
+    step_name = node.node_info.id
+    try:
+        return next(step for step in steps if step.name == step_name)
+    except StopIteration:
+        raise RuntimeError(f"Unable to find step with name '{step_name}'.")
+
+
+##### CONTEXT - RELATED UTILITIES #####
+def add_context_to_node(
+    pipeline_node: PipelineNode,
+    type_: str,
+    name: str,
+    properties: Dict[str, str],
+) -> None:
+    """
+    Add a new context to a TFX protobuf pipeline node.
+    Args:
+        pipeline_node: A tfx protobuf pipeline node
+        type_: The type name for the context to be added
+        name: Unique key for the context
+        properties: dictionary of strings as properties of the context
+    """
+    # Add a new context to the pipeline
+    context: ContextSpec = pipeline_node.contexts.contexts.add()
+    # Adding the type of context
+    context.type.name = type_
+    # Setting the name of the context
+    context.name.field_value.string_value = name
+    # Setting the properties of the context depending on attribute type
+    for key, value in properties.items():
+        c_property = context.properties[key]
+        c_property.field_value.string_value = value
+
+
+def serialize_pydantic_object(
+    obj: BaseModel, *, skip_errors: bool = False
+) -> Dict[str, str]:
+    """Convert a pydantic object to a dict of strings"""
+
+    class PydanticEncoder(json.JSONEncoder):
+        def default(self, o: Any) -> Any:
+            try:
+                return cast(Callable[[Any], str], obj.__json_encoder__)(o)
+            except TypeError:
+                return super().default(o)
+
+    def _inner_generator(
+        dictionary: Dict[str, Any]
+    ) -> Iterator[Tuple[str, str]]:
+        """Itemwise serialize each element in a dictionary."""
+        for key, item in dictionary.items():
+            try:
+                yield key, json.dumps(item, cls=PydanticEncoder)
+            except TypeError as e:
+                if skip_errors:
+                    logging.info(
+                        "Skipping adding field '%s' to metadata context as "
+                        "it cannot be serialized due to %s.",
+                        key,
+                        e,
+                    )
+                else:
+                    raise TypeError(
+                        f"Invalid type {type(item)} for key {key} can not be "
+                        "serialized."
+                    ) from e
+
+    return {key: value for key, value in _inner_generator(obj.dict())}
+
+
+def add_runtime_configuration_to_node(
+    pipeline_node: PipelineNode,
+    runtime_config: RuntimeConfiguration,
+) -> None:
```

Comment:
Saw a few methods that modified PipelineNode objects - may be worth separating these methods into their own file